checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -43,104 +92,621 @@ public class MustCallInferenceLogic {
+      Set<Obligation> obligations,
+      MethodInvocationNode mNode,
+      List<? extends VariableTree> paramsOfEnclosingMethod) {
+
     Node receiver = mNode.getTarget().getReceiver();
     if (receiver.getTree() == null) {
       return;
     }
 
     Element receiverEl = TreeUtils.elementFromTree(receiver.getTree());
+    if (receiverEl != null) {
+      if (receiverEl.getKind().isField()) {
+        inferOwningField(receiver, mNode);
+        return;
+      }
+    }
+
+    Node receiverTmpVar = mcca.getTempVarOrNode(receiver);
+    if (!(receiverTmpVar instanceof LocalVariableNode)) {
+      return;
+    }
+
+    Obligation receiverObligation =
+        MustCallConsistencyAnalyzer.getObligationForVar(
+            obligations, (LocalVariableNode) receiverTmpVar);
+    if (receiverObligation == null) {
+      return;
+    }
+
+    Set<ResourceAlias> receiverAliases = receiverObligation.resourceAliases;
+    for (int i = 0; i < paramsOfEnclosingMethod.size(); i++) {
+      VariableElement paramElt = TreeUtils.elementFromDeclaration(paramsOfEnclosingMethod.get(i));
 
-    if (receiverEl != null && typeFactory.isCandidateOwningField(receiverEl)) {
-      Element method = TreeUtils.elementFromUse(mNode.getTree());
-      List<String> mustCallValues = typeFactory.getMustCallValue(receiverEl);
+      for (ResourceAlias resourceAlias : receiverAliases) {
+        Element resourceElt = resourceAlias.reference.getElement();
+        if (!resourceElt.equals(paramElt)) {
+          continue;
+        }
+
+        JavaExpression target = JavaExpression.fromVariableTree(paramsOfEnclosingMethod.get(i));
+        if (mustCallObligationSatisfied(mNode, paramElt, target)) {
+          addOwningOnParams(i);
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks for indirect calls within the method represented by the given MethodInvocationNode. It
+   * checks the called-methods set of each argument after the call and adds owning to the field or
+   * parameter passed as an argument to this call.
+   *
+   * @param obligations Set of obligations associated with the current code block.
+   * @param mNode Method invocation node to check.
+   * @param paramsOfEnclosingMethod a list of the parameters of the enclosing method
+   */
+  private void checkIndirectCalls(
+      Set<Obligation> obligations,
+      MethodInvocationNode mNode,
+      List<? extends VariableTree> paramsOfEnclosingMethod) {
 
-      // This assumes that any MustCall annotation has at most one element.
+    List<Node> arguments = mcca.getArgumentsOfInvocation(mNode);
+    List<? extends VariableElement> parameters = mcca.getParametersOfInvocation(mNode);
+    if (parameters.isEmpty()) {
+      return;
+    }
+
+    for (int i = 0; i < arguments.size(); i++) {
+      Node arg = NodeUtils.removeCasts(arguments.get(i));
+      Element argElt = TreeUtils.elementFromTree(arg.getTree());
+      if (argElt == null && arg instanceof ArrayCreationNode) {
```

Comment:
I think this code block is inferring owning on the parameters of methods from calls to that method that pass an array of owning fields? I'm not 100% sure just from reading the code, so this code for arrays definitely needs an explanatory comment.

Reply:
This code block is to check the arguments passed in varArg positions with @EnsuresCalledMethodVarArgs annotation. I refactored this part of the code.